ezmlm-tstdig.1
.TH ezmlm-tstdig 1
.SH NAME
ezmlm-tstdig \- Tests if a digest should be created
.SH SYNOPSIS
.B ezmlm-tstdig
[
.B -k
.I kbytes
] [
.B -m
.I msg
] [
.B \-t
.I time
]
.I dir

.B ezmlm-tstdig
reads files in the list directory
.I dir
and determines if any of the criteria specified by the options are met.
If they are, or if all options specified are 0,
.B ezmlm-tstdig
exits with success (0). Otherwise,
.B ezmlm-tstdig
exits (99), or in case of errors, (100) for permanent and (111) for temporary
errors.

The normal use of
.B ezmlm-tstdig
is to place it in a script so that
.B ezmlm-get(1)
is executed if
.B ezmlm-tstdig
exits 0, but skipped otherwise.

.B ezmlm-tstdig
uses LOCAL to adapt its behavior so that it
can be used from the command line, a script, or from within
.IR dir\fB/editor ,
or
.IR dir\fB/manager .
If invoked from
.IR dir\fB/editor ,
the
.B ezmlm-tstdig
line should end in '|| exit 99'.
When used here,
.B ezmlm-tstdig
if digest generation criteria are met
tests a
timestamp in
.IR dir\fB/tstdig .
If
.I dir\fB/tstdig
is more than 1 hour old or a digest has been successfully created since
.I dir\fB/tstdig
was written,
.B ezmlm-tstdig
will set the timestamp to the current time and exit 0. Otherwise it will
exit 99. This is to guard against initiating duplicate digests when messages
arrive while digesting is in progress.
.SH OPTIONS
.TP
.B \-k
Success if more than
.I kbytes
of message body has accumulated since the latest digest.  Defaults to
the value in
.I dir\fB/digsize
if it exists or
.I 64
otherwise.
.TP
.B \-m
Success if more than
.I msg
messages have accumulated since the latest digest.  Defaults to the
value in
.I dir\fB/digcount
if it exists or
.I 30
otherwise.
.TP
.B \-t
Success if more than
.I time
hours have passed since the latest digest.  Defaults to the value in
.I dir\fB/digtime
if it exists or
.I 48
otherwise.
.SH USAGE
This is an example script for ezmlm-tstdig usage.

.RS
.nf
#!/bin/sh
#usage: script dir diglist@host
#digest if > 64 kbytes message bodies,
#          > 30 messages, or
#          > 48 hours since last digest.

# ezmlm bin path - modify for your setup
EZPATH='/usr/local/bin/ezmlm'

EZTST="${EZPATH}/ezmlm-tstdig"
EZGET="${EZPATH}/ezmlm-get"
${EZTST} -k64 -m30 -t48 "$1" || exit 0
${EZGET} -t "'$2'" "'$1'" < /dev/null
# improve by testing exit status
.fi
.RE
.SH BUGS
Obviously,
.B ezmlm-tstdig
should be integrated into the digesting program so that 
.I dir\fB/tstdig
will not be not necessary. This may be done in the next version.
.SH "SEE ALSO"
ezmlm-get(1),
ezmlm-make(1),
ezmlm-manage(1),
ezmlm-send(1),
ezmlm(5)


CHANGES
BUGS:	-ezmlm-idx does not implement a full rfc822 parser. Comments can be
	 inserted to defeat the MIME parsing mechanism, such as:
	 Content-Type: (comment) test/plain
	 This type of comment, while legal, does not occur in normal messages.
	 Guiding info past ezmlm-idx MIME parsing is easier done by enclosing
	 a message within a message. The mechanism isn't intended as a
	 "secure" way of preventing e.g. "application/ms-tnef" from being
	 disseminated.
	 If you think this is a problem, E-mail ezmlm@list.cr.yp.to.
	 Mimeremove/reject work only at the outer MIME layer.
	-ezmlm-request requires addresses to be user@host. While rfc822
	 compliant, addresses like ``"the man" @ host'' are not accepted.
	-ezmlm-cgi doesn't deal optimally with holes in the archive.
	-ezmlm-cgi seems not to work with Chinese. The problem has been to
	 find someone willing/able to participate in debugging.

Some ezmlmrc are not updated and may not work right.  If you fix them,
please send them to ezmlm@list.cr.yp.to for inclusion in the next
release.

ezmlm-idx-7.0.2, ????-??-??
===========================

- Fixed selection of [un]subscribe subject line for moderated
  subscriptions.

ezmlm-idx-7.0.1, 2009-07-09
===========================

- Added support for decoding sender addresses mangled by BATV prvs.

- Added missing ezmlm-checksub to the installed programs.

- Improved the TXT_BY message used in digests to work better with
  non-English languages.

- Fixed out-of-memory resulting in a permanent error.

- Fixed a buffer underflow bug in concatHDR.

- Fixed a bug in ezmlm-issubn.c that was causing it to scan the main
  list when it shouldn't.

- Fixed missing defaults in ezmlm-tstdig when digsize, digcount, or
  digtime were not created.

- Clarified the steps necessary when upgrading in UPGRADE.

ezmlm-idx-7.0.0, 2008-06-16
===========================

This version has three major architectural changes:

1. There is now support for internationalized messages.  All error
messages, all subject lines, and a few other messages are now configured
in a file named "text/messages" that is read at run time.  Messages in
this file will undergo header-style substitution before they are output.

Unlike other text files, all 3 of the files (DIR/text/messages,
/etc/ezmlm/LANG/text/messages, and /etc/ezmlm/default/text/messages) are
read, and only the first match (in the above order) is used for any
given message.  This allows for creation of partial files to override
just select messages.  If no match is found, the internal English text
is used.

2. The use of the "flags" file has been deprecated completely by
individual flag files accessed by the appropriate programs.  This, along
with corresponding changes to the ezmlmrc template, allows for lists to
be reconfigured without invoking ezmlm-make.  To upgrade a list to the
new setup, simply run "ezmlm-make -+ DIR".  Note that this will delete
the "config" and "flags" files, as they have been superceded by other
files.

3. Creation of subscriber tables has been moved out of ezmlm-mktab-* and
into ezmlm-make by adding some additional hooks to the subdb plugins.  A
ezmlm-rmtab program is also added to remove subscriber tables, also with
hooks in the subdb plugins.  The ezmlm-mktab-* programs have been
obsoleted by these changes and have been removed.

Smaller changes:

- Fixed a bug in ezmlm-request that caused a segfault when attempting to
  execute the "which" command.

- Added new program ezmlm-checksub to replace the
	ezmlm-issubn ... || { echo error; exit 100; }
  lines in the ezmlmrc files.  ezmlm-issubn is now deprecated for use in
  .qmail files, but still exists for backwards compatibility.

- Added ezmlm-weed to the confirmer control files.

- Added <#a#> substitution for the local part of the accept address.

- ezmlm-warn will now process bounces for both the main and digest lists
  if neither the -d nor the -D options are given.

- Added several control files for programs:
	dir/digcount		ezmlm-tstdig
	dir/digestcode		ezmlm-get
	dir/digformat		ezmlm-get
	dir/digsize		ezmlm-tstdig
	dir/digtime		ezmlm-tstdig
	dir/modcanedit		ezmlm-manage
	dir/modcanlist		ezmlm-manage
	dir/modgetonly		ezmlm-get
	dir/modpostonly		ezmlm-store
	dir/nosubconfirm	ezmlm-manage
	dir/nounsubconfirm	ezmlm-manage
	dir/nowarn		ezmlm-warn
	dir/subgetonly		ezmlm-get
	dir/noreturnposts	ezmlm-clean

- Removed the ezmlm-make -4 option in favor of the above files.

- Fixed handling of digest bouncer in ezmlm-dispatch.

- Eliminated the use of vfork for portability.

- Added support for decoding sender addresses mangled by SRS.

- Fixed bug when running ezmlm-archive on a newly-created list.

- Always enable ezmlm-request in manager.

- Always enable the "deny" blacklist.

- Fixed a long standing bug in ezmlm-store that caused both posting and
  moderating to happen when both modpost and confirmpost were disabled.

Note: The subdb API was modified in this version.  You will need to
reinstall all sub-* modules along with the main package.  Also, the
messages changes also replaced the confirmpost subject line hack in
ezmlm-store and the mailinglist file.  The contents of DIR/confirmpost
and DIR/mailinglist are now ignored in favor of the SUB_CONFIRM_POST and
TXT_MAILING_LIST messages, respectively.

ezmlm-idx-6.0.1, 2007-10-06
===========================

- Fixed all of the subdb plugins to correct a problem that prevented
  ezmlm-manage from working properly.

- Fixed ezmlm-weed to handle MIME Delivery Status Notification messages
  better.

- Fixed ezmlm-send to insert the proper value for the List-ID: header.
  Thanks Bill Nugent.

- Added the necessary Sender: header to make DomainKeys work into the
  ezmlmrc template.

- ezmlm-mktab-mysql and ezmlm-mktab-pgsql will now only be installed if
  they were built (with "make mysql" or "make pgsql" respectively).

Note: The subdb API was modified in this version.  You will need to
reinstall all sub-* modules along with the main package.


ezmlm-tstdig.c
#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include <unistd.h>
#include "stralloc.h"
#include "substdio.h"
#include "readwrite.h"
#include "strerr.h"
#include "sig.h"
#include "getconf.h"
#include "env.h"
#include "fmt.h"
#include "now.h"
#include "lock.h"
#include "getconfopt.h"
#include "messages.h"
#include "scan.h"
#include "case.h"
#include "str.h"
#include "open.h"
#include "die.h"
#include "wrap.h"
#include "idx.h"
#include "config.h"

const char FATAL[] = "ezmlm-tstdig: fatal: ";
const char USAGE[] =
"ezmlm-tstdig: usage: ezmlm-tstdig [-k kbytes] [-m messages] [-t hours] dir";

static unsigned long deltanum = ~0UL;
static unsigned long deltawhen = ~0UL;
static unsigned long deltasize = ~0UL;
static struct option options[] = {
  OPT_ULONG(deltasize,'k',"digsize"),
  OPT_ULONG(deltanum,'m',"digcount"),
  OPT_ULONG(deltawhen,'t',"digtime"),
  OPT_END
};

stralloc line = {0};

substdio ssnum;
char numbuf[16];

char strnum[FMT_ULONG];

int flaglocal = 0;

void main(int argc,char **argv)
{
  char *local;
  char *def;
  unsigned int pos;
  unsigned long num, digsize, dignum;
  unsigned long cumsize = 0L;
  unsigned long when, tsttime, digwhen;
  int fd,fdlock;

  (void) umask(022);
  sig_pipeignore();
  when = (unsigned long) now();
  optind = getconfopt(argc,argv,options,1,0);
  if (argv[optind])
    die_usage();	/* avoid common error of putting options after dir */
  if (!getconf_ulong2(&num,&cumsize,"num",0))
    _exit(99);

  if (getconf_line(&line,"dignum",0)) {
    if(!stralloc_0(&line)) die_nomem();
    pos = scan_ulong(line.s,&dignum);
    if (line.s[pos] == ':')
      pos += 1 + scan_ulong(line.s+pos+1,&digsize);
    if (line.s[pos] == ':')
      scan_ulong(line.s+pos+1,&digwhen);
  } else {
    dignum = 0L;	/* no file, not done any digest */
    digsize = 0L;	/* nothing digested */
    digwhen = 0L;	/* will force a digest, but the last one was eons  */
			/* ago. ezmlm-get sends it out only if there are   */
			/* messages. This is as it should for new lists.   */
  }
  local = env_get("LOCAL");
  if (local && *local) {			/* in editor or manager */
    def = env_get("DEFAULT");
    if (def && *def) {				/* in manager */
      if (!case_starts(def,"dig") || case_starts(def,"digest-"))
        _exit(0);
    } else {					/* in editor */
      flaglocal = 1;
    }
  }

  if (deltawhen == ~0UL)
    deltawhen = 48;
  if (deltasize == ~0UL)
    deltasize = 64;
  if (deltanum == ~0UL)
    deltanum = 30;
  if ((deltawhen && ((digwhen + deltawhen * 3600L) <= when)) ||
      (deltasize && ((digsize + (deltasize << 2)) <= cumsize)) ||
      (deltanum && ((dignum + deltanum) <= num))) {	/* digest! */
    if (flaglocal) {	/* avoid multiple digests. Of course, ezmlm-tstdig*/
			/* belongs in ezmlm-digest, but it's too late ....*/
      fdlock = lockfile("lock");
      getconf_line(&line,"tstdig",0);
      if (!stralloc_0(&line)) die_nomem();
      scan_ulong(line.s,&tsttime);	/* give digest 1 h to complete */
					/* nobody does digests more often */
      if ((tsttime + 3600L < when) || (tsttime <= digwhen)) {
        fd = open_trunc("tstdign");
        if (fd == -1)
          strerr_die2sys(111,FATAL,MSG1(ERR_CREATE,"tstdign"));
        substdio_fdbuf(&ssnum,write,fd,numbuf,sizeof(numbuf));
        if (substdio_put(&ssnum,strnum,fmt_ulong(strnum,when)) == -1)
          strerr_die2sys(111,FATAL,MSG1(ERR_WRITE,"tstdign"));
        if (substdio_puts(&ssnum,"\n") == -1)
          strerr_die2sys(111,FATAL,MSG1(ERR_WRITE,"tstdign"));
        if (substdio_flush(&ssnum) == -1)
          strerr_die2sys(111,FATAL,MSG1(ERR_FLUSH,"tstdign"));
        if (fsync(fd) == -1)
          strerr_die2sys(111,FATAL,MSG1(ERR_SYNC,"tstdign"));
        if (close(fd) == -1)
          strerr_die2sys(111,FATAL,MSG1(ERR_CLOSE,"tstdign"));
        wrap_rename("tstdign","tstdig");
        _exit(0);
      }
    } else
        _exit(0);
  }
  _exit(99);
}

